tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.cc

```diff
@@ -211,29 +270,136 @@ static StatusOr<bool> PadForTensorCores(HloCustomCallInstruction* conv) {
+  {
+    auto pad_dim = [](Shape* s, int64 dim) {
+      s->set_dimensions(dim, RoundUpToNearest<int64>(s->dimensions(dim), 4));
+    };
+
+    switch (kind) {
+      case CudnnConvKind::kForward:
+        CHECK(new_input_shapes.size() == 2);
+        pad_dim(&new_input_shapes[0],
+                dnums.input_feature_dimension());  // Input feature maps
+        pad_dim(&new_input_shapes[1],
+                dnums.kernel_input_feature_dimension());  // Kernel for the
+                                                          // input feature maps
+        pad_dim(
+            &new_input_shapes[1],
+            dnums.kernel_output_feature_dimension());  // Kernel for the output
+                                                       // feature maps
+        pad_dim(&new_result_shape,
+                dnums.output_feature_dimension());  // Output feature maps
+        break;
+      case CudnnConvKind::kForwardActivation:
+        CHECK(new_input_shapes.size() == 3 || new_input_shapes.size() == 4);
+        pad_dim(&new_input_shapes[0],
+                dnums.input_feature_dimension());  // Input feature maps
+        pad_dim(&new_input_shapes[1],
+                dnums.kernel_input_feature_dimension());  // Kernel for the
+                                                          // input feature maps
+        pad_dim(
+            &new_input_shapes[1],
+            dnums.kernel_output_feature_dimension());  // Kernel for the output
+                                                       // feature maps
+        pad_dim(&new_input_shapes[2], 0);              // Bias
+        if (new_input_shapes.size() == 4) {
+          pad_dim(&new_input_shapes[3],
+                  dnums.output_feature_dimension());  // Optional side input
+        }
+        pad_dim(&new_result_shape,
+                dnums.output_feature_dimension());  // Output feature maps
+        break;
+      default:
+        CHECK(false);
     }
+    // Check that padding wouldn't increase the total bytes read/written by this
+    // operation too much.
+    auto check_size_increase = [&](const Shape& old_shape,
+                                   const Shape& new_shape) {
+      int64 old_bytes = ShapeUtil::ByteSizeOf(old_shape);
+      int64 new_bytes = ShapeUtil::ByteSizeOf(new_shape);
+      if (new_bytes <= old_bytes * kMaxBytesTouchedIncrease) {
+        return;
+      }
+      VLOG(3)
+          << "Not padding convolution; doing so would change input / result "
+             "shape from "
+          << ShapeUtil::HumanString(old_shape) << " to "
+          << ShapeUtil::HumanString(new_shape) << ", a size increase of "
+          << new_bytes / static_cast<double>(old_bytes) << "x > "
+          << kMaxBytesTouchedIncrease << "x: " << conv->ToString();
+      return;
+    };
+
+    for (int64 i = 0; i < conv->operand_count(); ++i) {
+      check_size_increase(conv->operand(i)->shape(), new_input_shapes[i]);
+    }
+    check_size_increase(result_shape, new_result_shape);
   }
```

Comment:
This is kind of a bad bug.  This lambda doesn't return anything.  Therefore in the lines below, we don't check anything.  And apparently there are no unit tests that covered this case.

This went undetected for ~9 months.